### DISCUSSION TITLE
How to trigger a Dagster job from an endpoint call in Python?

### DISCUSSION THREAD

**Discussion by yuval-airis** [dagster-io/dagster#18245] [Q&A] [answered]
I want to run a webserver that would allow trigger dagster jobs.

Both the webserver (Say, [uvicorn](https://www.uvicorn.org/)) and dagster run on their own processes, so I can't just call [job.execute_in_process](https://docs.dagster.io/_apidocs/execution).

What alternatives do I have?
How may I trigger a dagster job from another process?

**Comment by yuval-airis** [comment]
Ok, I noticed two possible solutions for that:
[job-execution#command-line](https://docs.dagster.io/concepts/ops-jobs-graphs/job-execution#command-line)
[GraphQL execution](https://docs.dagster.io/concepts/webserver/graphql-client)

Which may allow the webserver to call the dagster service and trigger a new job

**Comment by lvzhengcq** [comment, marked as answer]
To trigger a Dagster job from another process, particularly from a web server like Uvicorn, you can use Dagster's DagsterGraphQLClient to send a GraphQL request to the Dagster instance, triggering the desired pipeline run.
Here's a basic example of how you might achieve this:
from dagster import DagsterGraphQLClient

# Set up the GraphQL client to communicate with your Dagster instance
dagster_client = DagsterGraphQLClient('http://localhost:3000/graphql') # Replace with your Dagster instance URL

# Define a function to trigger a Dagster job
def trigger_dagster_job(pipeline_name, environment_dict):
try:
response = dagster_client.execute_plan(
pipeline_name=pipeline_name,
environment_dict=environment_dict,
)
# Process the response if needed
print(response)
return response
except Exception as e:
# Handle exceptions
print(f"Error triggering job: {e}")
return None

# Call the function to trigger the Dagster job
result = trigger_dagster_job('your_pipeline_name', {'config_key': 'config_value'})

Replace 'http://localhost:3000/graphql' with the actual URL where your Dagster instance is running. Also, replace 'your_pipeline_name' with the name of the pipeline you want to trigger, and provide the appropriate environment configuration in the form of a dictionary as needed by your pipeline.
Ensure that your web server (Uvicorn) has appropriate permissions and network access to communicate with the Dagster instance.
This method uses Dagster's GraphQL API to send a request to start a pipeline run. Make sure that your Dagster instance is configured correctly to receive these requests and execute the specified pipeline.
Remember, when triggering jobs remotely, security considerations are important. Ensure that you implement appropriate authentication and authorization mechanisms to prevent unauthorized access to trigger Dagster jobs.

**Reply by yuval-airis** [reply]
Thanks!
What is the pipeline name? I have a series of assets (each an op really) with dependencies which creates a kind of graph.

Just to test this I created another flow with a job, and used the name of the job "do_stuff" (Basically the name of the function with the @job decorator and it failed too, with this error:
`>>> dagster_client.submit_job_execution("do_stuff")
Traceback (most recent call last):
  File "/home/ubuntu/.local/share/virtualenvs/code-gF5vZsWE/lib/python3.11/site-packages/urllib3/connection.py", line 174, in _new_conn
    conn = connection.create_connection(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/share/virtualenvs/code-gF5vZsWE/lib/python3.11/site-packages/urllib3/util/connection.py", line 72, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/socket.py", line 961, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
socket.gaierror: [Errno -3] Temporary failure in name resolution

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/.local/share/virtualenvs/code-gF5vZsWE/lib/python3.11/site-packages/urllib3/connectionpool.py", line 715, in urlopen
    httplib_response = self._make_request(
                       ^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/share/virtualenvs/code-gF5vZsWE/lib/python3.11/site-packages/urllib3/connectionpool.py", line 416, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/ubuntu/.local/share/virtualenvs/code-gF5vZsWE/lib/python3.11/site-packages/urllib3/connection.py", line 244, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/usr/lib/python3.11/http/client.py", line 1282, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.11/http/client.py", line 1328, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.11/http/client.py", line 1277, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.11/http/client.py", line 1037, in _send_output
    self.send(msg)
  File "/usr/lib/python3.11/http/client.py", line 975, in send
    self.connect()
  File "/home/ubuntu/.local/share/virtualenvs/code-gF5vZsWE/lib/python3.11/site-packages/urllib3/connection.py", line 205, in connect
    conn = self._new_conn()
           ^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/share/virtualenvs/code-gF5vZsWE/lib/python3.11/site-packages/urllib3/connection.py", line 186, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7ffaeae95fd0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution`

What could this mean?
